### ISSUE TITLE
Require user to explicitly detach() when doing "a += b"

### ISSUE THREAD

**Issue by jaak-s** [pytorch/pytorch#24961]
## 🚀 Feature
For a person not knowing the internals of Pytorch, the semantics of `a += b` are not fully obvious when both `a` and `b` require gradients. The proposal is to:
1. Raise an error when `a += b` is executed when both require gradients.
2. The error message should recommend to choose between `a += b.detach()` and `a = a + b`.

Thus, the idea is to require users to **explicitly** detach `b`, to obtain the current behavior.

## Motivation
While intending to execute `a = a + b` it is easy to make a mistake due out of habit or not paying attention and write `a += b` . Since there is no error and warning, the code silently drops the gradients to `b` which usually just results in lower model accuracy. Thus, the error is quite hard to catch, without explicitly searching the code base for `+=`.

## Pitch
Raise an error when `a += b` is called when both `a` and `b` have `requires_grad == True`. Error message should explain the two options 1) `a += b.detach()` for executing the current behavior, 2) `a = a + b` for keeping the gradient.

Same for other operators, like `a *= b`, `a -= b` etc.

## Alternatives
Alternative would be to raise a warning when `a += b` which would serve the same purpose but without breaking code where `a += b` was meant to do `a += b.detach()`.

**Comment by rsokl** [comment]
PyTorch does not silently drop the gradient for `b` on in-place operations.

```python
>>> import torch as tr
>>> a = tr.tensor(1., requires_grad=True)                                                                             
>>> b = tr.tensor(2., requires_grad=True)                                                                             
>>> c = 2*a                                                                                                           
>>> c *= b                                                                                                            
>>> c.backward()                                                                                                      
>>> a.grad, b.grad                                                                                                    
>>> (tensor(4.), tensor(2.))
```
see that `b` does indeed have the appropriate gradient, even though it was introduced to the computational graph via augmented update.

**Comment by jaak-s** [comment]
You are right!
I stumbled on the comments in bug #9634 that give an opposite explanation. I'll close this bug for now.
